experimental/oak_async/tests/fake_runtime.rs

```diff
@@ -0,0 +1,130 @@
+    let handles =
+        unsafe { core::slice::from_raw_parts_mut(buf as *mut HandleWithStatus, count as usize) };
+    WAIT_ON_CHANNELS_HANDLER.with(|handler| {
+        (handler
+            .borrow_mut()
+            .as_mut()
+            .expect("no wait_on_channels handler configured"))(handles)
+    }) as i32 as u32
+}
+
+type WaitOnChannelsHandler = Box<dyn FnMut(&mut [HandleWithStatus]) -> OakStatus>;
+std::thread_local! {
+    static WAIT_ON_CHANNELS_HANDLER: RefCell<Option<WaitOnChannelsHandler>> = RefCell::new(None);
+}
+
+pub fn set_wait_on_channels_handler<F: 'static + FnMut(&mut [HandleWithStatus]) -> OakStatus>(
+    handler: F,
+) {
+    WAIT_ON_CHANNELS_HANDLER.with(|cell| cell.replace(Some(Box::new(handler))));
+}
+
+#[no_mangle]
+pub extern "C" fn channel_read(
+    handle: Handle,
+    buf: *mut u8,
+    size: usize,
+    actual_size: *mut u32,
+    _handle_buf: *mut u8,
+    _handle_count: u32,
+    actual_handle_count: *mut u32,
+) -> u32 {
+    READY_DATA.with(|ready_data| {
+        let mut ready_data = ready_data.borrow_mut();
+        let data_queue = ready_data.entry(handle).or_default();
+        let status = match data_queue.front() {
+            None => OakStatus::ErrChannelEmpty,
+            Some(Err(status)) => *status,
+            Some(Ok(data)) if data.len() > size => {
+                unsafe { *actual_size = data.len() as u32 };
+                OakStatus::ErrBufferTooSmall
+            }
+            Some(Ok(data)) => {
+                unsafe {
+                    let buf = core::slice::from_raw_parts_mut(buf, data.len());
+                    buf.copy_from_slice(&data);
+                    *actual_size = data.len() as u32;
+                    *actual_handle_count = 0;
+                };
+                let _ = data_queue.pop_front();
+                OakStatus::Ok
+            }
+        };
+        status as i32 as u32
+    })
+}
+
+type ReadyChannelData = VecDeque<Result<Vec<u8>, OakStatus>>;
+
+std::thread_local! {
+    static READY_DATA: RefCell<HashMap<Handle, ReadyChannelData>> = RefCell::new(HashMap::new());
+}
+
+pub fn add_ready_data<T: Encodable>(handle: Handle, data: &T) {
+    let msg = data.encode().expect("Failed to encode ready data");
+    READY_DATA.with(|ready_data| {
+        ready_data
+            .borrow_mut()
+            .entry(handle)
+            .or_default()
+            .push_back(Ok(msg.bytes))
+    });
+}
+
+// Note: this status is added into the queue of ready data, so any previously add ready data must
+// first be read before this status will be returned. The channel status is never removed, any
+// future reads will always return this error status.
+pub fn set_error(handle: Handle, status: OakStatus) {
+    assert_ne!(status, OakStatus::Ok);
+    READY_DATA.with(|ready_data| {
+        ready_data
+            .borrow_mut()
+            .entry(handle)
+            .or_default()
+            .push_back(Err(status))
+    });
+}
```

Comment:
Could you add perhaps one high-level comment to one of these items (or to the package) to explain how it's meant to be used? No need to document all the individual fns if they are self-explanatory after the high-level comment.

Reply:
That's a good idea, done!